### PR TITLE
#98 Add 'WithParents'

### PR DESCRIPTION
**Pull request by IvanKirpichnikov**

### CODE REVIEW

File: tests/unit/container/test_with_parents.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+from dishka.exceptions import NoFactoryError
+
+
+def test_simple_inheritance() -> None:
+    class A1: ...
+    class A2(A1): ...
+    class A3(A2): ...
+    
+    provider = Provider(scope=Scope.APP)
+    provider.provide(lambda: 1, provides=WithParents[A3])
```
Comment (Tishka17) [inline tests/unit/container/test_with_parents.py:15]: let's check that value generates only once when accessing by multiple classes, so as there are aliases

File: tests/unit/container/test_with_parents.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    container = make_container(provider)
+    assert container.get(A3) == 1
+    assert container.get(A2) == 1
+    assert container.get(A1) == 1
+
+
+def test_ignore_parent_type() -> None:
```
Comment (Tishka17) [inline tests/unit/container/test_with_parents.py:22]: parametrize

File: tests/unit/container/test_with_parents.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    provider = Provider(scope=Scope.APP)
+    provider.provide(lambda: 1, provides=WithParents[A2])
+    
+    container = make_container(provider)
+    
+    assert container.get(A2) == 1
+    assert container.get(A1[str, int, type]) == 1
+
+
+def test_type_var_and_type_var_tuple() -> None:
```
Comment (Tishka17) [inline tests/unit/container/test_with_parents.py:84]: parametrize?

File: src/dishka/entities/with_parents.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        )
+    else:
+        parents = get_parents_for_mro(obj)
+    return parents
+
+
+if TYPE_CHECKING:
+    from typing import Union as WithParents
+else:
+    class WithParents:
```
Comment (Tishka17) [inline src/dishka/entities/with_parents.py:185]: add to package `__init__.py`

File: .ruff.toml [hunk 5]
```diff
@@ -24,6 +24,8 @@ lint.ignore = [
    "PLR0913",
    "UP038",
    "TCH001",
+   "SIM103",
+   "PLR1704",
```
Comment (Tishka17) [inline .ruff.toml:28]: not this probably

Comment (IvanKirpichnikov) [reply .ruff.toml]: done

File: src/dishka/entities/with_parents.py (hunk 4 above)
Comment (IvanKirpichnikov) [reply src/dishka/entities/with_parents.py:177]: done

File: tests/unit/container/test_with_parents.py (hunk 2 above)
Comment (IvanKirpichnikov) [reply tests/unit/container/test_with_parents.py]: done

File: tests/unit/container/test_with_parents.py (hunk 1 above)
Comment (IvanKirpichnikov) [reply tests/unit/container/test_with_parents.py]: done

File: tests/unit/container/test_with_parents.py (hunk 3 above)
Comment (IvanKirpichnikov) [reply tests/unit/container/test_with_parents.py:80]: not done

Comment (IvanKirpichnikov) [reply tests/unit/container/test_with_parents.py]: done

File: src/dishka/entities/with_parents.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    return hasattr(obj, "__orig_bases__")
+
+if HAS_PY_311:
+    def is_type_var_tuple(obj: TypeHint) -> bool:
+        return getattr(obj, "__typing_is_unpacked_typevartuple__", False)
+else:
+    def is_type_var_tuple(obj: TypeHint) -> bool:
+        return False
+
+def is_ignore_type(origin_obj: TypeHint) -> bool:
```
Comment (Tishka17) [inline src/dishka/entities/with_parents.py:47]: is_ignored_type

Comment (IvanKirpichnikov) [reply src/dishka/entities/with_parents.py]: done

File: tests/unit/container/test_with_parents.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+def test_ignore_parent_type(obj: Any, value: Any, component: Any) -> None:
+    provider = Provider(scope=Scope.APP, component=component)
+    provider.provide(
+        source=lambda: obj(),
+        provides=Annotated[WithParents[obj], FromComponent(component)],
+    )
+    container = make_container(provider)
+
+    try:
+        container.get(value, component=component)
```
Comment (Tishka17) [inline tests/unit/container/test_with_parents.py:59]: pytest.raises

**Pull request merged**